python/cudf/cudf/core/column/column.py

```diff
@@ -80,21 +80,23 @@ def data_array_view(self):
         else:
             dtype = self.dtype
 
-        result = rmm.device_array_from_ptr(
-            ptr=self.data.ptr, nelem=len(self), dtype=dtype
+        result = cuda.as_cuda_array(self.data)
```

Comment:
Does this hold a reference to `self.data` now? It didn't used to which is why we stuck in the attribute below.

Reply:
Yeah RMM's `device_array_from_ptr` does not hold a reference as it doesn't know what object it came from.

However Numba's `as_cuda_array` does. From [the docs]( http://numba.pydata.org/numba-doc/latest/cuda/memory.html )...

> A view of the underlying GPU buffer is created. No copying of the data is done. The resulting DeviceNDArray will acquire a reference from `obj`.

This can also be verified in the code itself. The function [`as_cuda_array` calls `from_cuda_array_interface`]( https://github.com/numba/numba/blob/d7953a18dbf5ea231dc16e967ce8e9b754578ea6/numba/cuda/api.py#L66-L67 ), which in turn [builds a `MemoryPointer` with a reference to `obj` (in this context called `owner`)]( https://github.com/numba/numba/blob/d7953a18dbf5ea231dc16e967ce8e9b754578ea6/numba/cuda/api.py#L50 ) used by the [`DeviceNDArray`]( https://github.com/numba/numba/blob/d7953a18dbf5ea231dc16e967ce8e9b754578ea6/numba/cuda/api.py#L51-L52 ) returned.

IOW the resulting `DeviceNDArray`, will hold a reference to our `Buffer` object (`self.data` in this case) as long as that array is in scope.

Reply:
Should add in the special case of `data_array_view`. We actually toss the first `DeviceNDArray` we are given, but hang onto the underling `MemoryPointer` in order to build a new `DeviceNDArray`. As the `MemoryPointer` is what owns the reference to `self.data` and we do pass this to the new `DeviceNDArray`, it will remain valid as long as the new `DeviceNDArray` is in scope.

Just to explain why this workaround is need, using `DeviceNDArray`'s `.view(...)` with a type that has a different itemsize than what we started with, is currently not supported by Numba. ( https://github.com/numba/numba/issues/4829 ) So this works around that issue by simply building a new `DeviceNDArray`. A fix in flight ( https://github.com/numba/numba/pull/5085 ) so should hopefully be unneeded in a future version of Numba.

This workaround is unneeded in `mask_array_view` as `uint8` and `int8` have the same itemsize. So we can just call `.view(...)`.

Reply:
@jakirkham I think this was a somewhat recent change to `as_cuda_array` as it didn't used to hold a reference and we don't pin the absolute latest version of numba. Any chance you can confirm this behavior with the oldest version of Numba we support?

Reply:
Looks like it goes all the way back to `0.44.0` so we're probably good. cc @shwina for visibility

Reply:
Ah I see what you mean. Thanks for that info 🙂

Looking back it appears commit ( https://github.com/numba/numba/commit/6d4c6f9fca4ff5af030010f590f853b327269237 ) added tracking the `owner`. This seems to be part of PR ( https://github.com/numba/numba/pull/2860 ), which also published the `__cuda_array_interface__` spec as part of Numba 0.39.0. So this seems pretty safe. Though of course there could have been bugs in-between, which I may not be aware of.